### PR TITLE
Reduce Boost dependency footprint

### PR DESCRIPTION
**Pull request by andrjohns**
## Summary

Including Boost headers outside of their Math library tends to pull in a surprising amount of other Boost headers.

For example, for the single `lexical_cast` call in `init_threadpool_tbb()` the total Boost footprint is:

  - Boost.Config
  - Boost.TypeTraits
  - Boost.Container 
  - Boost.Core
  - Boost.Detail
  - Boost.Exception
  - Boost.ThrowException
  - Boost.Limits
  - Boost.Move
  - Boost.StaticAssert
  - Boost.Version
  - Boost.LexicalCast 

This PR removes three different types of Boost dependencies:

  1. `boost/lexical_cast` (`init_threadpool_tbb`): Replaced with a `std::from_chars` + try/catch
  2. `boost/optional` (`hypergeometric_2F1`): Replaced with `std::optional`
  3. `boost/random` (`_rng` functions): Replaced with `std::` equivalents (e.g., `std::bernoulli_distribution bernoulli_rng`)

## Tests

N/A - Existing tests should still pass.

Note that this triggered an edge-case with the expected-value (chi-square) testing for distributions, where too few values were sampled for particular bins. The testing util was updated to just merge bins if the expected counts were too low.

## Side Effects

Possibility of changes to existing `_rng` outputs with same seed value

## Release notes

Replaced Boost dependencies with standard library equivalents

## Checklist

- [x] Copyright holder: Andrew Johnson

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Coding-Style-and-Idioms) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by WardBrian** [comment]
Funny, I was just talking to someone today about decreasing our reliance on boost/random

Can we add any new libraries to the trim list
https://github.com/stan-dev/math/blob/cc6ba8bcf31996d581d67c887653009df3b9d7dd/lib/upgrade-boost.sh#L117-L120
after this?

File: test/unit/math/prim/prob/neg_binomial_test.cpp [hunk 1]
```diff
@@ -178,6 +177,30 @@ TEST(ProbDistributionsNegBinomial, chiSquareGoodnessFitTest3) {
     bin[i] = 0;
   expected_bin_sizes(expect, K, N, alpha, beta);
 
+  // alpha < 1 makes this distribution's tail very sparse, so most of the K
+  // bins have expected counts far below N; a single sample landing in one
+  // of them can dominate the statistic. Pool bins so each group has an
+  // expected count of at least 5, per standard chi-square goodness-of-fit
+  // practice. Grouping is based only on the theoretical expected counts
+  // (not the realized samples), so it doesn't bias the test.
```
Comment (SteveBronder) [inline test/unit/math/prim/prob/neg_binomial_test.cpp:185]: You can add this hook to your claude and it stops it from doing these multi line comments

https://gist.github.com/SteveBronder/8c2f141fed4511447a8ad5939157857e

Generally I have seen that if you let it start writing a lot of these it uses them as an excuse to do bad code. The comment can also be wrong since it is untested which can lead to other agents down the road getting confused here.

**Review by SteveBronder — DISMISSED** [review]
Cool!

**Comment by SteveBronder** [comment]
From the tests it seems like the changes in `stan` have to happen first before we can merge this

File: test/unit/math/prim/prob/neg_binomial_test.cpp (hunk 1 above)
Comment (andrjohns) [reply test/unit/math/prim/prob/neg_binomial_test.cpp:185]: Oh super cool, immediately stolen

**Comment by andrjohns** [comment]
Given the discovery of the issues with `<random>` distributions in https://github.com/stan-dev/stan/pull/3412, I'm going to close this one and open a separate PR without those changes
